From: ZhangPeng <zhangpeng362@huawei.com>
Subject: mm: oom: terminate the oom_evaluate_task() loop early
Date: Mon, 14 Aug 2023 14:34:28 +0800

If task is allocating a lot of memory and has been marked to be killed
first, then it gets the highest score (LONG_MAX).  Therefore, there is no
need to continue to calculate the points of other tasks.  Just terminate
the oom_evaluate_task() loop early, when the task with the highest score
is found.  By doing this, we can get some performance gains in
select_bad_process().

To implement it, the return value of oom_evaluate_task() is modified. 
When the task with the highest score is found (points == LONG_MAX),
oom_evaluate_task() will return 1 and the loop will terminate early.

Link: https://lkml.kernel.org/r/20230814063428.4111206-3-zhangpeng362@huawei.com
Signed-off-by: ZhangPeng <zhangpeng362@huawei.com>
Cc: Kefeng Wang <wangkefeng.wang@huawei.com>
Cc: Michal Hocko <mhocko@suse.com>
Cc: Nanyong Sun <sunnanyong@huawei.com>
Cc: Shakeel Butt <shakeelb@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/oom_kill.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/mm/oom_kill.c~mm-oom-terminate-the-oom_evaluate_task-loop-early
+++ a/mm/oom_kill.c
@@ -346,6 +346,8 @@ static int oom_evaluate_task(struct task
 	get_task_struct(task);
 	oc->chosen = task;
 	oc->chosen_points = points;
+	if (points == LONG_MAX)
+		return 1;
 next:
 	return 0;
 abort:
_
